src/vs/workbench/api/common/extHostTesting.ts

```diff
@@ -222,6 +228,159 @@ export class ExtHostTesting implements ExtHostTestingShape {
+		});
+
+		return {
+			root: TestItemFilteredWrapper.getWrapperForTestItem(workspaceHierarchy.root, document.uri),
+			dispose: () => {
+				onDidChangeTest.dispose();
+				TestItemFilteredWrapper.removeFilter(document.uri);
+			},
+			onDidChangeTest: onDidChangeTest.event
+		};
+	}
+}
+
+/*
+ * A class which wraps a vscode.TestItem that provides the ability to filter a TestItem's children
+ * to only the children that are located in a certain vscode.Uri.
+ */
+class TestItemFilteredWrapper implements vscode.TestItem {
+	private static wrapperMap = new WeakMap<vscode.Uri, WeakMap<vscode.TestItem, TestItemFilteredWrapper>>();
+	public static removeFilter(uri: vscode.Uri): void {
+		this.wrapperMap.delete(uri);
+	}
+
+	// Wraps the TestItem specified in a TestItemFilteredWrapper and pulls from a cache if it already exists.
+	public static getWrapperForTestItem(item: vscode.TestItem, filterToUri: vscode.Uri, parent?: TestItemFilteredWrapper): TestItemFilteredWrapper {
+		let innerMap = this.wrapperMap.get(filterToUri);
+		if (innerMap?.has(item)) {
+			return innerMap.get(item)!;
+		}
+
+		if (!innerMap) {
+			innerMap = new WeakMap<vscode.TestItem, TestItemFilteredWrapper>();
+			this.wrapperMap.set(filterToUri, innerMap);
+
+		}
+
+		const w = new TestItemFilteredWrapper(item, filterToUri, parent);
+		innerMap.set(item, w);
+		return w;
+	}
+
+	public get label() {
+		return this.actual.label;
+	}
+
+	public get debuggable() {
+		return this.actual.debuggable;
+	}
+
+	public get description() {
+		return this.actual.description;
+	}
+
+	public get location() {
+		return this.actual.location;
+	}
+
+	public get runnable() {
+		return this.actual.runnable;
+	}
+
+	public get state() {
+		return this.actual.state;
+	}
+
+	public get children() {
+		// We only want children that match the filter.
+		return this.getWrappedChildren().filter(child => child.hasNodeMatchingFilter);
+	}
+
+	public get visibleParent(): TestItemFilteredWrapper {
+		return this.hasNodeMatchingFilter ? this : this.parent!.visibleParent;
+	}
+
+	private matchesFilter: boolean | undefined;
+
+	// Determines if the TestItem matches the filter. This would be true if:
+	// 1. We don't have a parent (because the root is the workspace root node)
+	// 2. The URI of the current node matches the filter URI
+	// 3. Some child of the current node matches the filter URI
+	public get hasNodeMatchingFilter(): boolean {
+		if (this.matchesFilter === undefined) {
+			this.matchesFilter = !this.parent
+				|| this.actual.location?.uri.toString() === this.filterToUri.toString()
+				|| this.getWrappedChildren().some(child => child.hasNodeMatchingFilter);
+		}
+
+		return this.matchesFilter;
+	}
+
+	// Reset the cache of whether or not you can see a node from a particular node
+	// up to it's visible parent.
+	public reset(): void {
+		if (this === this.visibleParent) {
+			this.matchesFilter = undefined;
+		} else {
+			this.parent?.reset();
+			this.matchesFilter = undefined;
+		}
```

Comment:
nit: can be simplified slightly

```suggestion
		if (this !== this.visibleParent) {
			this.parent?.reset();
		}
		this.matchesFilter = undefined;
```

Reply:
fixed